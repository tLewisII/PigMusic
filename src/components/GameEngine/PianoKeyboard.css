.piano-keyboard {
  background: #8B4513;
  padding: 20px;
  border-radius: 20px;
  box-shadow: 0 8px 24px rgba(0, 0, 0, 0.2);
}

.keys-container {
  display: flex;
  position: relative;
  gap: 4px;
}

.key-wrapper {
  position: relative;
}

.white-key {
  width: 60px;
  height: 200px;
  background: white;
  border: 2px solid #ddd;
  border-radius: 0 0 8px 8px;
  cursor: pointer;
  position: relative;
  transition: all 0.1s ease;
  display: flex;
  align-items: flex-end;
  justify-content: center;
  padding-bottom: 20px;
  font-weight: bold;
  font-size: 1.2rem;
  color: #333;
}

.white-key:hover:not(.disabled) {
  background: #f5f5f5;
  transform: translateY(2px);
}

.white-key:active:not(.disabled) {
  background: #e0e0e0;
  transform: translateY(4px);
}

.white-key.disabled {
  background: #eee;
  color: #ccc;
  cursor: not-allowed;
  opacity: 0.5;
}

.white-key.highlighted {
  background: #FFB6C1;
  animation: glow 1s ease-in-out infinite;
}

.black-key {
  position: absolute;
  width: 40px;
  height: 120px;
  background: #333;
  border: none;
  border-radius: 0 0 6px 6px;
  cursor: pointer;
  top: 0;
  right: -20px;
  z-index: 1;
  transition: all 0.1s ease;
  display: flex;
  align-items: flex-end;
  justify-content: center;
  padding-bottom: 15px;
  color: white;
  font-size: 0.9rem;
  font-weight: bold;
}

.black-key:hover:not(.disabled) {
  background: #444;
  transform: translateY(2px);
}

.black-key:active:not(.disabled) {
  background: #555;
  transform: translateY(4px);
}

.black-key.disabled {
  background: #666;
  cursor: not-allowed;
  opacity: 0.3;
}

.black-key.highlighted {
  background: #FF69B4;
  animation: glow 1s ease-in-out infinite;
}

.key-label {
  user-select: none;
}

@keyframes glow {
  0%, 100% {
    box-shadow: 0 0 10px rgba(255, 105, 180, 0.5);
  }
  50% {
    box-shadow: 0 0 20px rgba(255, 105, 180, 0.8);
  }
}

/* Mobile responsive styles */
@media (max-width: 768px) {
  .piano-keyboard {
    padding: 12px;
    border-radius: 12px;
  }

  .keys-container {
    gap: 3px;
    justify-content: center;
  }

  .white-key {
    width: 38px;
    height: 120px;
    padding-bottom: 12px;
    font-size: 0.85rem;
    border-radius: 0 0 6px 6px;
  }

  .black-key {
    width: 26px;
    height: 75px;
    right: -13px;
    padding-bottom: 8px;
    font-size: 0.7rem;
    border-radius: 0 0 4px 4px;
  }
}