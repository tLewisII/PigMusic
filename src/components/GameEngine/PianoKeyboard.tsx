import React from 'react';
import AudioManager from '../../services/AudioManager';
import './PianoKeyboard.css';

interface PianoKeyboardProps {
  availableNotes: string[];
  onNoteSelect: (note: string) => void;
  highlightedNote?: string;
}

const PianoKeyboard: React.FC<PianoKeyboardProps> = ({
  availableNotes,
  onNoteSelect,
  highlightedNote
}) => {
  const whiteKeys = ['C', 'D', 'E', 'F', 'G', 'A', 'B', 'C', 'D', 'E', 'F'];
  const blackKeys = ['C#', 'D#', null, 'F#', 'G#', 'A#', null, 'C#', 'D#', null];
  
  const isNoteAvailable = (note: string, octave: string = '4'): boolean => {
    return availableNotes.includes(note + octave);
  };

  const getNoteName = (note: string, index: number): string => {
    // After B (index 6), we move to octave 5
    const octave = index <= 6 ? '4' : '5';
    return note + octave;
  };

  const handleKeyPress = async (note: string) => {
    // Extract note name and octave
    const noteName = note.slice(0, -1);
    const octave = note.slice(-1);
    
    if (isNoteAvailable(noteName, octave)) {
      await AudioManager.playNote(note);
      onNoteSelect(note);
    }
  };

  return (
    <div className="piano-keyboard">
      <div className="keys-container">
        {whiteKeys.map((note, index) => {
          const fullNote = getNoteName(note, index);
          const octave = index <= 6 ? '4' : '5';
          const available = isNoteAvailable(note, octave);
          const highlighted = highlightedNote === fullNote;
          
          return (
            <div key={index} className="key-wrapper">
              <button
                className={`white-key ${!available ? 'disabled' : ''} ${highlighted ? 'highlighted' : ''}`}
                onClick={() => handleKeyPress(fullNote)}
                disabled={!available}
              >
                <span className="key-label">{note}{octave}</span>
              </button>
              
              {index < blackKeys.length && blackKeys[index] && (
                <button
                  className={`black-key ${!isNoteAvailable(blackKeys[index]!, octave) ? 'disabled' : ''} ${highlightedNote === blackKeys[index] + octave ? 'highlighted' : ''}`}
                  onClick={() => handleKeyPress(blackKeys[index] + octave)}
                  disabled={!isNoteAvailable(blackKeys[index]!, octave)}
                >
                  <span className="key-label">{blackKeys[index]}</span>
                </button>
              )}
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default PianoKeyboard;