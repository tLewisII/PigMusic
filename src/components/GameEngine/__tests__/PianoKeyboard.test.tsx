import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PianoKeyboard from '../PianoKeyboard';

describe('PianoKeyboard', () => {
  const mockOnKeyPress = jest.fn();

  beforeEach(() => {
    mockOnKeyPress.mockClear();
  });

  it('renders correctly without hint', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with hint key', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} hintKey={40} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correct number of keys', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} />
    );
    
    // Should have 52 white keys (C to C across 7+ octaves)
    const whiteKeys = container.querySelectorAll('.key:not(.black)');
    expect(whiteKeys).toHaveLength(52);
    
    // Should have 36 black keys
    const blackKeys = container.querySelectorAll('.key.black');
    expect(blackKeys).toHaveLength(36);
  });

  it('calls onKeyPress with correct key number when white key is clicked', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} />
    );
    
    // Click middle C (key 40)
    const middleC = container.querySelectorAll('.key:not(.black)')[20]; // 21st white key
    fireEvent.click(middleC);
    
    expect(mockOnKeyPress).toHaveBeenCalledWith(40);
  });

  it('calls onKeyPress with correct key number when black key is clicked', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} />
    );
    
    // Click C# (key 41)
    const cSharp = container.querySelectorAll('.key.black')[14]; // 15th black key
    fireEvent.click(cSharp);
    
    expect(mockOnKeyPress).toHaveBeenCalledWith(41);
  });

  it('highlights hint key with special class', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} hintKey={40} />
    );
    
    const hintedKey = container.querySelector('[data-key="40"]');
    expect(hintedKey).toHaveClass('hint');
  });

  it('displays note labels on white keys', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} />
    );
    
    // Check for note labels
    expect(container.textContent).toContain('C');
    expect(container.textContent).toContain('D');
    expect(container.textContent).toContain('E');
    expect(container.textContent).toContain('F');
    expect(container.textContent).toContain('G');
    expect(container.textContent).toContain('A');
    expect(container.textContent).toContain('B');
  });

  it('renders keyboard with proper structure', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} />
    );
    
    expect(container.querySelector('.piano-keyboard')).toBeInTheDocument();
    expect(container.querySelector('.keys-container')).toBeInTheDocument();
  });

  it('prevents event propagation on black key clicks', () => {
    const { container } = render(
      <PianoKeyboard onKeyPress={mockOnKeyPress} />
    );
    
    const blackKey = container.querySelector('.key.black');
    const clickEvent = new MouseEvent('click', { bubbles: true });
    const stopPropagationSpy = jest.spyOn(clickEvent, 'stopPropagation');
    
    blackKey?.dispatchEvent(clickEvent);
    
    expect(stopPropagationSpy).toHaveBeenCalled();
  });
});