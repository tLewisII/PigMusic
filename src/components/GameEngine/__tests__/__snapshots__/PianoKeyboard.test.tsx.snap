// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PianoKeyboard renders correctly with highlighted note 1`] = `
<div
  class="piano-keyboard"
>
  <div
    class="keys-container"
  >
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  highlighted"
      >
        <span
          class="key-label"
        >
          C
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          C#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          D
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          D#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          E
          4
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          F
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          F#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          G
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          G#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          A
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          A#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          B
          4
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`PianoKeyboard renders correctly without hint 1`] = `
<div
  class="piano-keyboard"
>
  <div
    class="keys-container"
  >
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          C
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          C#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          D
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          D#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          E
          4
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          F
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          F#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key  "
      >
        <span
          class="key-label"
        >
          G
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          G#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          A
          4
        </span>
      </button>
      <button
        class="black-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          A#
        </span>
      </button>
    </div>
    <div
      class="key-wrapper"
    >
      <button
        class="white-key disabled "
        disabled=""
      >
        <span
          class="key-label"
        >
          B
          4
        </span>
      </button>
    </div>
  </div>
</div>
`;
